File: backend/app/schemas/analysis.py
from typing import Dict, Any, Optional
from pydantic import BaseModel

class MissingValuesSummary(BaseModel):
    count: int
    percentage: float

class NumericalStats(BaseModel):
    count: float
    mean: Optional[float]
    std: Optional[float]
    min: Optional[float]
    p25: Optional[float]
    p50: Optional[float]
    p75: Optional[float]
    max: Optional[float]

class CategoricalStats(BaseModel):
    unique_count: int
    top_values: Dict[str, int]

class AnalysisResponse(BaseModel):
    dataset_id: str
    row_count: int
    col_count: int
    data_types: Dict[str, str]
    missing_values: Dict[str, MissingValuesSummary]
    duplicate_rows: int
    numerical_stats: Dict[str, NumericalStats]
    categorical_stats: Dict[str, CategoricalStats]
    correlation_matrix: Dict[str, Dict[str, Optional[float]]]

class DatasetPreviewResponse(BaseModel):
    columns: list[str]
    rows: list[Dict[str, Any]]
    total_rows: int
    page: int
    limit: int

class ChartQueryRequest(BaseModel):
    x_axis: str
    y_axis: Optional[str] = None
    chart_type: str  # "bar", "line", "pie", "histogram", "box", "scatter"
    agg_func: Optional[str] = None  # "sum", "mean", "count", "min", "max"
    group_by: Optional[str] = None

class ChartQueryResponse(BaseModel):
    labels: list[Any]
    values: list[Any]
    chart_type: str
    x_axis_label: str
    y_axis_label: str

File: frontend/src/api/analysis.ts
import { apiClient } from './client';

export interface MissingValuesSummary {
  count: number;
  percentage: number;
}

export interface NumericalStats {
  count: number;
  mean: number | null;
  std: number | null;
  min: number | null;
  p25: number | null;
  p50: number | null;
  p75: number | null;
  max: number | null;
}

export interface CategoricalStats {
  unique_count: number;
  top_values: Record<string, number>;
}

export interface AnalysisResponse {
  dataset_id: string;
  row_count: number;
  col_count: number;
  data_types: Record<string, string>;
  missing_values: Record<string, MissingValuesSummary>;
  duplicate_rows: number;
  numerical_stats: Record<string, NumericalStats>;
  categorical_stats: Record<string, CategoricalStats>;
  correlation_matrix: Record<string, Record<string, number | null>>;
}

export interface DatasetPreviewResponse {
  columns: string[];
  rows: Record<string, any>[];
  total_rows: number;
  page: number;
  limit: number;
}

export interface ChartQueryRequest {
  x_axis: string;
  y_axis?: string;
  chart_type: string;
  agg_func?: string;
  group_by?: string;
}

export interface ChartQueryResponse {
  labels: any[];
  values: any[];
  chart_type: string;
  x_axis_label: string;
  y_axis_label: string;
}

export const getDatasetSummary = async (datasetId: string): Promise<AnalysisResponse> => {
  const response = await apiClient.get<AnalysisResponse>(`/analysis/${datasetId}/summary`);
  return response.data;
};

export const getDatasetPreview = async (
  datasetId: string,
  page: number = 1,
  limit: number = 20,
  search?: string
): Promise<DatasetPreviewResponse> => {
  const params = new URLSearchParams({
    page: page.toString(),
    limit: limit.toString(),
  });
  if (search) params.append('search', search);

  const response = await apiClient.get<DatasetPreviewResponse>(`/analysis/${datasetId}/preview?${params.toString()}`);
  return response.data;
};

export const queryChartData = async (
  datasetId: string,
  request: ChartQueryRequest
): Promise<ChartQueryResponse> => {
  const response = await apiClient.post<ChartQueryResponse>(`/analysis/${datasetId}/query`, request);
  return response.data;
};
